src/docs/markdown/extending-caddy/placeholders.md

```diff
@@ -0,0 +1,140 @@
+
+If you are not yet familiar with placeholders, start by reading [here](/docs/conventions#placeholders)!
+
+## Placeholder Internals
+
+Internally, placeholders are simply a string in format `{foo.bar}` used as valid configuration values, which is later parsed at runtime.
+
+Placeholders-like strings which start with a dollar sign (`{$FOO}`), are evaulated at Caddyfile parse time, and do not need to be dealt with by your plugin. This is because these are not placeholders, but Caddyfile-specific [environmental variable substitution](/docs/caddyfile/concepts#environment-variables), they just happen to share the `{}` syntax.
+
+It is therefore important to understand that `{env.HOST}` is inherently different from something like `{$HOST}`
+
+As an example, see the following caddyfile:
+```caddyfile
+:8080 {
+  respond {$HOST} 200
+}
+
+:8081 {
+  respond {env.HOST} 200
+}
+```
+
+When you adapt this Caddyfile with `HOST=example caddy adapt` you will get
+```json
+{
+  "apps": {
+    "http": {
+      "servers": {
+        "srv0": {
+          "listen": [
+            ":8080"
+          ],
+          "routes": [
+            {
+              "handle": [
+                {
+                  "body": "example",
+                  "handler": "static_response",
+                  "status_code": 200
+                }
+              ]
+            }
+          ]
+        },
+        "srv1": {
+          "listen": [
+            ":8081"
+          ],
+          "routes": [
+            {
+              "handle": [
+                {
+                  "body": "{env.HOST}",
+                  "handler": "static_response",
+                  "status_code": 200
+                }
+              ]
+            }
+          ]
+        }
+      }
+    }
+  }
+}
+```
+
+Importantly, look at the `"body"` field in both `srv0` and `srv1`.
+
+Since `srv0` used `{$HOST}`, the special environmental variable replacement with `$`, the value became `example`, as it was processed during Caddyfile parse time.
+
+Since `srv1` used `{env.HOST}`, a normal placeholder, it was parsed as its own raw string value, `{env.HOST}`
+
+Some users may immediately notice that this means it is impossible to use the `{$ENV}` syntax in a JSON config. The solution to this is to process such placeholders at Provision time, which is covered below.
+
+
+## How to use placeholders in your plugin
+
+#### Parse the raw placeholder value in your unmarshaler
+
+Placeholders are not evaluated at Caddyfile parse time, and should be preserved for later use. They are used as their raw string values.
+
+In other words, parsing a placeholder is no different from parsing any other string.
+
+```go
+func (g *Gizmo) UnmarshalCaddyfile(d *caddyfile.Dispenser) error {
+	d.Next()
+	if !d.Args(&g.Name) {
+		// not enough args
+		return d.ArgErr()
+	}
+}
+```
+
+#### Evaluate the placeholder during Match or Serve
```

Comment:
I would reframe this section and the next to be more general.

It boils down to whether a Replacer is already available or not. If one has already been made, as is the case with the HTTP handler chain, it can be obtained from the context. I think we can talk about it with that as the example, but less specific to "Match" or "ServeHTTP". The only real difference is how you get the Replacer: either from a context, or make a new one.

Reply:
do you think it should be reframed to be more general, or there be instead two articles,

1. explaining how to grab a new replacer, and using that to resolve a placeholder
2. explaining how to create a replacer from a context.Context.

Reply:
I think this one article is enough; mainly let's show examples of both in this one.

So show how to get it from the context, where it will often be stored, and also how to create a new one if there isn't one already available.